src/types/module.rs

```diff
@@ -388,8 +349,294 @@ impl PyModule {
+    /// ```text
+    /// Hello world!
+    /// ```
+    ///
+    /// [1]: crate::prelude::pyfunction
+    /// [2]: crate::wrap_pyfunction
+    fn add_function(&self, fun: &Bound<'_, PyCFunction>) -> PyResult<()>;
+}
+
+impl<'py> PyModuleMethods<'py> for Bound<'py, PyModule> {
+    fn dict(&self) -> Bound<'py, PyDict> {
+        unsafe {
+            // PyModule_GetDict returns borrowed ptr; must make owned for safety (see #890).
+            ffi::PyModule_GetDict(self.as_ptr())
+                .assume_borrowed(self.py())
+                .to_owned()
+                .downcast_into_unchecked()
+        }
+    }
+
+    fn index(&self) -> PyResult<Bound<'py, PyList>> {
+        let __all__ = __all__(self.py());
+        match self.getattr(__all__) {
+            Ok(idx) => idx.downcast_into().map_err(PyErr::from),
+            Err(err) => {
+                if err.is_instance_of::<exceptions::PyAttributeError>(self.py()) {
+                    let l = PyList::empty(self.py()).as_borrowed().to_owned();
+                    self.setattr(__all__, &l).map_err(PyErr::from)?;
+                    Ok(l)
+                } else {
+                    Err(err)
+                }
+            }
+        }
+    }
+
+    fn name(&self) -> PyResult<&str> {
+        self.as_borrowed().name()
+    }
+
+    #[cfg(not(PyPy))]
+    fn filename(&self) -> PyResult<Bound<'py, PyString>> {
+        use crate::py_result_ext::PyResultExt;
+
+        unsafe {
+            ffi::PyModule_GetFilenameObject(self.as_ptr())
+                .assume_owned_or_err(self.py())
+                .downcast_into_unchecked()
+        }
+    }
+
+    fn add<V>(&self, name: &str, value: V) -> PyResult<()>
+    where
+        V: IntoPy<PyObject>,
+    {
+        self.index()?
+            .append(name)
+            .expect("could not append __name__ to __all__");
+        self.setattr(name, value.into_py(self.py()))
+    }
+
+    fn add_class<T>(&self) -> PyResult<()>
+    where
+        T: PyClass,
+    {
+        let py = self.py();
+        self.add(T::NAME, T::lazy_type_object().get_or_try_init(py)?)
+    }
+
+    fn add_wrapped<T>(&self, wrapper: &impl Fn(Python<'py>) -> T) -> PyResult<()>
+    where
+        T: IntoPyCallbackOutput<PyObject>,
+    {
+        fn inner(module: &Bound<'_, PyModule>, object: Bound<'_, PyAny>) -> PyResult<()> {
+            let name = object.getattr(__name__(module.py()))?;
+            module.add(name.extract()?, object)
+        }
+
+        let py = self.py();
+        inner(self, wrapper(py).convert(py)?.into_bound(py))
+    }
+
+    fn add_submodule(&self, module: &Bound<'_, PyModule>) -> PyResult<()> {
+        let name = module.name()?;
+        self.add(name, module)
+    }
+
+    fn add_function(&self, fun: &Bound<'_, PyCFunction>) -> PyResult<()> {
+        let name = fun.getattr(__name__(self.py()))?;
+        self.add(name.extract()?, fun)
+    }
+}
+
+impl<'a, 'py> Borrowed<'a, 'py, PyModule> {
+    fn name(self) -> PyResult<&'a str> {
```

Comment:
This works because the name of a module is immutable? If so, a comment might be nice.

Reply:
Great catch, it looks to me like this is not sound, the `__name__` is just stored in the module dictionary and this could easily produce a dangling reference. Will change this.

https://github.com/python/cpython/blob/3f84a19e6291db682fc9a570e7612e80e2ffbbb5/Objects/moduleobject.c#L540-L550